atlite/convert.py

```diff
@@ -119,22 +120,24 @@ def convert_and_aggregate(
 
     no_args = all(v is None for v in [layout, shapes, matrix])
 
+    results = None    
+
     if no_args:
         if per_unit or return_capacity:
             raise ValueError(
                 "One of `matrix`, `shapes` and `layout` must be "
                 "given for `per_unit` or `return_capacity`"
             )
-        if capacity_factor or capacity_factor_timeseries:
-            if capacity_factor_timeseries:
-                res = da.rename("capacity factor")
-            else:
-                res = da.mean("time").rename("capacity factor")
-            res.attrs["units"] = "p.u."
-            return maybe_progressbar(res, show_progress, **dask_kwargs)
+            
+    if capacity_factor or capacity_factor_timeseries:
+        if capacity_factor_timeseries:
+            results = da.rename("capacity factor")
         else:
-            res = da.sum("time", keep_attrs=True)
-            return maybe_progressbar(res, show_progress, **dask_kwargs)
+            results = da.mean("time").rename("capacity factor")
+        results.attrs["units"] = "p.u."
+
+    else:
+        results = da.sum("time", keep_attrs=True)
```

Comment:
Removing the return means that we no longer end the function here, which leads to undefined issues later. I know, this one is open for quite a while now, but could you have another look?

Otherwise this function is quite a mess and could get a general refactoring with your proposed changes included

Reply:
@lkstrp: I need to check it locally, but could you explain which undefined issues will occur later? I know the fix is not ideal, and was actually just done that kind of messy to not restructure the whole function.

Reply:
Further down, `matrix` is not defined. If you merge the current master, this will hopefully show up in the tests. They may not run through because of a missing cdsapi token, as this branch was derived earlier. But you can run them locally in any case using `pytest .`.

``` python
if index is None:
    index = pd.RangeIndex(matrix.shape[0])
```

But as you said, restructuring the whole function is better. If you have already looked at it, feel free. Otherwise I will deal with it at some point.

Reply:
@lkstrp I restructured the function and tested it. I commit it soon. Nevertheless, this would need to change some things in the origional test files, too and in the usage. I explain it with the commit, Best, Tim

Reply:
Sounds great! Thank you @TimFuermann

Reply:
@lkstrp Just uploaded a new version of how the convert and aggregate could be used more consistently. Nevertheless this includes heavy changes in the input flags, but allows for a more consistent way of using them. A default capacity layout is used if no capacities are provided. This is a constant 1MW per gridcell (which should be the same as looking for capacity factors). With this new implementation, capacity factors for regions, per gridcell, aswell as aggregated can be calculated.

Reply:
@TimFuermann Still haven't found time do dive deeper in this which I need to do. But I will get back to you!